### ISSUE TITLE
RangeError: Maximum call stack size exceeded when require/load very big json file

### ISSUE THREAD

**Issue by dongyuwei** [facebook/hermes#135]
https://github.com/dongyuwei/tiny_english_dictionary/blob/hermes/trie-service.js#L2 
`const words = require('./words_with_frequency_and_translation_and_ipa.json');` will throw error: 
```
RangeError: Maximum call stack size exceeded, js engine:hermes
```

The json file size is 21M, while its data structure is very simple.

### Steps to reproduce the bug:
1. yarn start
2. start an android emulator
3. yarn android

Tested with yarn 1.17.3 and nodejs 12.12.0 on Mac 10.14.5.

**Comment by dongyuwei** [comment]
Exception call stack:
![image](https://user-images.githubusercontent.com/112451/66893741-4a247100-f021-11e9-9be8-bd0b739812e1.png)

**Comment by tmikov** [comment]
Thanks for reporting this. It seems like we have a problem with compiling large JSON files. We are looking into it and will post an update here.

Meanwhile, as a workaround can you parse as a string with JSON.parse()?

**Comment by dongyuwei** [comment]
Thanks for your reply, I'll try and let you know today. On road now.

**Comment by dongyuwei** [comment]
I just verified JSON.parse works.

**Comment by sofiageo** [comment]
In my case I have a 14MB file which builds fine in debug mode but Hermes hangs when trying to optimize. I had to turn off hermes optimizations for the build to complete. Let me know if I can help to somehow debug this for you.

**Comment by dongyuwei** [comment]
There is a unit test to parse json: `/unittests/Parser/JSONParserTest.cpp`, but I don't know how to run the test. 

@tmikov can you modify and run the test with my  big json file `words_with_frequency_and_translation_and_ipa.json`? From the call stack of the exception, can't see any direct relationship with hermes.

**Comment by tmikov** [comment]
We investigated this when it was reported and we believe we understand the bug. We haven't forgotten it, and we have been workin to address it, but progress has been slow because the fix is is non-trivial. There are two separate problems:

- Without optimizations, the generated code for large nested object literals consume inordinate amount of register stack, causing the stack overflow exception.
- With optimizations, the stack usage is much better, but one particular optimization pass exhibits degenerate behavior on the same kind of object literals.

The workaround is to use `JSON.parse()` with a string or a resource.

I will look into escalating this, so the fix can land sooner.

**Comment by aprilmintacpineda** [comment]
This happened to me when I upgraded my react-native https://react-native-community.github.io/upgrade-helper/?from=0.63.4&to=0.64.0 

```
 WARN  Require cycle: node_modules/core-js/internals/microtask.js -> node_modules/core-js/internals/microtask.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/rn-fetch-blob/index.js -> node_modules/rn-fetch-blob/polyfill/index.js -> node_modules/rn-fetch-blob/polyfill/Blob.js -> node_modules/rn-fetch-blob/index.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/rn-fetch-blob/index.js -> node_modules/rn-fetch-blob/polyfill/index.js -> node_modules/rn-fetch-blob/polyfill/XMLHttpRequest.js -> node_modules/rn-fetch-blob/index.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 WARN  Require cycle: node_modules/rn-fetch-blob/index.js -> node_modules/rn-fetch-blob/polyfill/index.js -> node_modules/rn-fetch-blob/polyfill/Fetch.js -> node_modules/rn-fetch-blob/index.js

Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
 ERROR  RangeError: Maximum call stack size exceeded, js engine: hermes
 LOG  Running "ny_app" with {"rootTag":1}
 ERROR  Invariant Violation: "ny_app" has not been registered. This can happen if:
* Metro (the local dev server) is run from the wrong folder. Check if Metro is running, stop it and restart it in the current project.
* A module failed to load due to an error and `AppRegistry.registerComponent` wasn't called., js engine: hermes
```

**Comment by melroyd16** [comment]
> This happened to me when I upgraded my react-native https://react-native-community.github.io/upgrade-helper/?from=0.63.4&to=0.64.0
> 
> ```
>  WARN  Require cycle: node_modules/core-js/internals/microtask.js -> node_modules/core-js/internals/microtask.js
> 
> Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
>  WARN  Require cycle: node_modules/rn-fetch-blob/index.js -> node_modules/rn-fetch-blob/polyfill/index.js -> node_modules/rn-fetch-blob/polyfill/Blob.js -> node_modules/rn-fetch-blob/index.js
> 
> Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
>  WARN  Require cycle: node_modules/rn-fetch-blob/index.js -> node_modules/rn-fetch-blob/polyfill/index.js -> node_modules/rn-fetch-blob/polyfill/XMLHttpRequest.js -> node_modules/rn-fetch-blob/index.js
> 
> Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
>  WARN  Require cycle: node_modules/rn-fetch-blob/index.js -> node_modules/rn-fetch-blob/polyfill/index.js -> node_modules/rn-fetch-blob/polyfill/Fetch.js -> node_modules/rn-fetch-blob/index.js
> 
> Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
>  ERROR  RangeError: Maximum call stack size exceeded, js engine: hermes
>  LOG  Running "ny_app" with {"rootTag":1}
>  ERROR  Invariant Violation: "ny_app" has not been registered. This can happen if:
> * Metro (the local dev server) is run from the wrong folder. Check if Metro is running, stop it and restart it in the current project.
> * A module failed to load due to an error and `AppRegistry.registerComponent` wasn't called., js engine: hermes
> ```
> 
> Happens both on android and ios.

Setting inlineRequires: false in metro.config.js fixed it for me

**Comment by chensu-chime** [comment]
> We investigated this when it was reported and we believe we understand the bug. We haven't forgotten it, and we have been workin to address it, but progress has been slow because the fix is non-trivial. There are two separate problems:
> 
> * Without optimizations, the generated code for large nested object literals consume inordinate amount of register stack, causing the stack overflow exception.
> * With optimizations, the stack usage is much better, but one particular optimization pass exhibits degenerate behavior on the same kind of object literals.
> 
> The workaround is to use `JSON.parse()` with a string or a resource.
> 
> I will look into escalating this, so the fix can land sooner.

Is there any update on the fix?

**Comment by traingethermarc** [comment]
Having the same problem with a JSON file - no chance to use a big json... any help?

**Comment by traingethermarc** [comment]
I have a 13mb JSON file which should be loaded locally. I tried basically everything, but it always gives me this error... Can someone help me? Any working work-around for this? Or a fix somehow? Thank you...

**Comment by tmikov** [comment]
@traingethermarc the workaround is to use JSON.parse().

**Comment by aprilmintacpineda** [comment]
I'm getting this error but I don't even import any big json files, trying to get to the root cause.

**Comment by aprilmintacpineda** [comment]
> This happened to me when I upgraded my react-native https://react-native-community.github.io/upgrade-helper/?from=0.63.4&to=0.64.0
> 
> ```
>  WARN  Require cycle: node_modules/core-js/internals/microtask.js -> node_modules/core-js/internals/microtask.js
> 
> Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
>  WARN  Require cycle: node_modules/rn-fetch-blob/index.js -> node_modules/rn-fetch-blob/polyfill/index.js -> node_modules/rn-fetch-blob/polyfill/Blob.js -> node_modules/rn-fetch-blob/index.js
> 
> Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
>  WARN  Require cycle: node_modules/rn-fetch-blob/index.js -> node_modules/rn-fetch-blob/polyfill/index.js -> node_modules/rn-fetch-blob/polyfill/XMLHttpRequest.js -> node_modules/rn-fetch-blob/index.js
> 
> Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
>  WARN  Require cycle: node_modules/rn-fetch-blob/index.js -> node_modules/rn-fetch-blob/polyfill/index.js -> node_modules/rn-fetch-blob/polyfill/Fetch.js -> node_modules/rn-fetch-blob/index.js
> 
> Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
>  ERROR  RangeError: Maximum call stack size exceeded, js engine: hermes
>  LOG  Running "ny_app" with {"rootTag":1}
>  ERROR  Invariant Violation: "ny_app" has not been registered. This can happen if:
> * Metro (the local dev server) is run from the wrong folder. Check if Metro is running, stop it and restart it in the current project.
> * A module failed to load due to an error and `AppRegistry.registerComponent` wasn't called., js engine: hermes
> ```
> 
> Happens both on android and ios.

<div class="js-discussion js-socket-channel ml-0 pl-0 ml-md-6 pl-md-3" data-channel="eyJjIjoibWFya2VkLWFzLXJlYWQ6MjEwMzI0MTkiLCJ0IjoxNjI0MjM1NTg1fQ==--f21294cc2a1d85c0b23df0aed9361a6b94dc6259c4842f17e5d3bc044505ea64" data-channel-target="MDU6SXNzdWU1MDc2Mzc3MjY=" style="box-sizing: border-box; margin-left: 40px !important; padding-left: 16px !important; color: rgb(201, 209, 217); font-family: -apple-system, system-ui, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div style="box-sizing: border-box;"><div style="box-sizing: border-box;"><div class="js-timeline-item js-timeline-progressive-focus-container" data-gid="MDEyOklzc3VlQ29tbWVudDg2NDY0MDgzNA==" style="box-sizing: border-box;"><div class="TimelineItem js-comment-container" data-gid="MDEyOklzc3VlQ29tbWVudDg2NDY0MDgzNA==" data-url="/_render_node/MDEyOklzc3VlQ29tbWVudDg2NDY0MDgzNA==/timeline/issue_comment?variables%5BdeferredCommentActions%5D=false" style="box-sizing: border-box; position: relative; display: flex; padding: 16px 0px; margin-left: 16px;"><div class=" timeline-comment-group js-minimizable-comment-group js-targetable-element TimelineItem-body my-0 " id="issuecomment-864640834" style="box-sizing: border-box; margin-top: 4px; margin-bottom: 0px !important; min-width: 0px; max-width: 100%; color: var(--color-timeline-text); flex: 1 1 auto;"><div class="ml-n3 timeline-comment unminimized-comment comment previewable-edit js-task-list-container editable-comment js-comment timeline-comment--caret reorderable-task-lists current-user" data-body-version="ee0a6de2c179deb0f4916efaaec7f433713ba18efdf0b1281b1fc602576b1c9a" data-unfurl-hide-url="/content_reference_attachments/hide" style="box-sizing: border-box; margin-left: -16px !important; position: relative; color: var(--color-text-primary); background-color: var(--color-bg-primary); border-color: var(--color-box-border-info); border-top-style: ; border-top-width: ; border-right-style: ; border-right-width: ; border-bottom-style: ; border-bottom-width: ; border-left-style: ; border-left-width: ; border-image-source: ; border-image-slice: ; border-image-width: ; border-image-outset: ; border-image-repeat: ; border-radius: 6px;"><div class="edit-comment-hide" style="box-sizing: border-box;"><task-lists sortable="" style="box-sizing: border-box;">I'm getting this error but I don't even import any big json files, trying to get to the root cause.</task-lists></div><form class="js-comment-update" id="issuecomment-864640834-edit-form" action="https://github.com/facebook/hermes/issue_comments/864640834" accept-charset="UTF-8" method="post" style="box-sizing: border-box;"></form></div></div></div></div><div style="box-sizing: border-box;"><div class="js-timeline-marker js-socket-channel js-updatable-content" id="partial-timeline" data-channel="eyJjIjoiaXNzdWU6NTA3NjM3NzI2IiwidCI6MTYyNDIzNTcxMH0=--875ce9c7db9aa50deca3492702d948b72e86baafce5cbc49720f53fd3efcf4c2" data-url="/_render_node/MDU6SXNzdWU1MDc2Mzc3MjY=/issues/unread_timeline?variables%5BdeferCommitBadges%5D=false&amp;variables%5BdeferStatusCheckRollups%5D=false&amp;variables%5BdeferredCommentActions%5D=true&amp;variables%5BhasFocusedReviewComment%5D=false&amp;variables%5BhasFocusedReviewThread%5D=false&amp;variables%5BtimelinePageSize%5D=30&amp;variables%5BtimelineSince%5D=2021-06-21T00%3A35%3A07Z" data-last-modified="Mon, 21 Jun 2021 00:35:07 GMT" data-gid="MDU6SXNzdWU1MDc2Mzc3MjY=" style="box-sizing: border-box;"></div></div></div></div></div><span id="issue-comment-box" style="box-sizing: border-box; color: rgb(201, 209, 217); font-family: -apple-system, system-ui, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"></span><span style="color: rgb(201, 209, 217); font-family: -apple-system, system-ui, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(13, 17, 23); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial; display: inline !important; float: none;"></span><div class="discussion-timeline-actions" style="box-sizing: border-box; background-color: var(--color-bg-canvas); border-top: 2px solid var(--color-border-primary); color: rgb(201, 209, 217); font-family: -apple-system, system-ui, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="timeline-comment-wrapper timeline-new-comment js-comment-container js-targetable-element  ml-0 pl-0 ml-md-6 pl-md-3" id="issuecomment-new" style="box-sizing: border-box; margin-left: 40px !important; padding-left: 56px; position: relative; margin-top: 16px; margin-bottom: 0px;"><div class=" d-none d-md-block" style="box-sizing: border-box; display: block !important;"><br class="Apple-interchange-newline"></div></div></div>

**Comment by hainguyentien** [comment]
> @traingethermarc the workaround is to use JSON.parse().

Can you tell me more how to handle it with JSON.parse()? Because I import the json, using JSON.parse(JSON.stringify()) but it not work

**Comment by dongdyang** [comment]
it has been two years. And this issue still occurs for the latest 0.9 version hermes.

**Comment by cristianoccazinsp** [comment]
What's hermes' max call stack size? I'm getting max call stack size errors for recursive functions as deep as just 60 calls.

**Comment by tmikov** [comment]
@cristianoccazinsp For JS to JS recursion, the maximum depth is dynamic and depends on the size of every frame and the size of the register stack, which is configurable. In the default configuration, the following code performs 55,000 recursive invocations without optimization, and 87,000 with optimization:
```
let level = 0;
function foo() {
    ++level;
    foo();
}

try {
    foo();
} catch (e) {
    print("Level", level);
    print(e);
}
```

Hermes has a separate fixed stack limit for native calls, which is set to 128. So, if you change the recursive invocation from `foo()` to `foo.call()`, it throws after 128 recursive calls.

**Comment by cristianoccazinsp** [comment]
Native stack errors is just the crash I'm facing right now. Sigh, 128 is so small! This used to work fine with JSC and now blows up with Hermes.  Sadly, the code is for an HTML parsing library that relies on recursion so 128 will definitely cause issues as HTMLs could be huge.

**Comment by tmikov** [comment]
Keep in mind that the low number is only for cases involving recursion through a native function. The number is deliberately very conservative to guarantee that a native stack overflow will never occur. It is even lower in some build modes, because the native stack (which is not controlled by Hermes) can and does overflow quite easily.

The limit can be increased by changing this line:
https://github.com/facebook/hermes/blob/2c66a23882bb7b6c9f80fadd84df643e73bbd5ee/include/hermes/VM/Runtime.h#L905

There is some work in progress for obtaining and checking the actual stack size, which on most cases will result in significantly increased depth, but no promises when it will land.

**Comment by cristianoccazinsp** [comment]
I'm not entirely sure why I get native stack depth errors even though the recursion happens entirely in JS. Does calls to `.map` and "native" methods count towards the native stack?

**Comment by tmikov** [comment]
Yes. Most "standard library" functions are implemented in C++ and unfortunately count as native.

The best way to address this is to check whether the native stack is actually overflowing, instead of setting artificial hard limits. I will see if I can finish that work.

**Comment by cristianoccazinsp** [comment]
@tmikov I replaced all `.map` calls with simple `for a in b` loops and the recursion errors went away (even though I'm still using recursive calls). Funny thing is, the moment the screen gets unmounted, another native stack depth error occurs from a totally different place (react-navigation stack and animations).

**Comment by aikewoody** [comment]
With this comment I want to raise awareness of this issue and show that there is a high demand for it to be fixed. (Even though there is a workaround to import as string and then use `JSON.parse`)

**Comment by kdthomas2121** [comment]
Can you detail how to use the workaround @aikewoody? I'm still getting this crash when using require("pathToJson") before I get the change to parse it

**Comment by aikewoody** [comment]
@kdthomas2121
To be honest I don't fully remember what the solution was exactly.

One trick that I used is giving the json file a `.txt` extension so that it wouldn't be automatically parsed.

This is the code I ended up with (React Native + Expo):
`JSON.parse(await FileSystem.readAsStringAsync('~/assets/data/list-production.txt'));`

**Comment by Hau-Hau** [comment]
Same issue occurs for big JS/TS file.

**Comment by rsteel1** [comment]
> @kdthomas2121 To be honest I don't fully remember what the solution was exactly.
> 
> One trick that I used is giving the json file a `.txt` extension so that it wouldn't be automatically parsed.
> 
> This is the code I ended up with (React Native + Expo): `JSON.parse(await FileSystem.readAsStringAsync('~/assets/data/list-production.txt'));`

Every chance you don't remember again but does this require putting the json file into the device local storage?

Also adding a comment to ask for a fix to this.

**Comment by mikeduminy** [comment]
@tmikov how can we tell if `HERMES_LIMIT_STACK_DEPTH` is defined in the build we are using? https://github.com/facebook/hermes/blob/main/include/hermes/VM/Runtime.h#L918-L921

**Comment by tmikov** [comment]
@mikeduminy sorry, I have lost the context here. What kind of error are you getting and when?

**Comment by mikeduminy** [comment]
@tmikov 

**Context**
We have a large react-native app that uses code splitting to lazily load code as needed. Locally we use Metro which doesn't support code splitting -- it transforms all dynamic imports into regular `require`s. This causes the whole app to be a single, large bundle. We also use the Hermes runtime in development (this ensures we catch any engine-specific differences). 

**Problem**
When this large bundle (100+ mb) is loaded the runtime throws an error:
```RangeError: Maximum call stack size exceeded, js engine: hermes```

**Investigation**
The only way we're able to make it work again (with metro + Hermes runtime) is by commenting out a bunch of these dynamic requires which are causing the bundle to grow so much. Importantly, by commenting them out at random, we are able to get it running again. This led us to believe that it's probably not about specific parts of our code but rather the sheer quantity of code.

The reason I asked about `HERMES_LIMIT_STACK_DEPTH` is because in it can cause `MAX_NATIVE_CALL_FRAME_DEPTH` to be [very low](https://github.com/facebook/hermes/blob/main/include/hermes/VM/Runtime.h#L918-L921). I'd like to not have to adjust internal limits within hermes to fix the problem but we're getting desperate.

**Comment by tmikov** [comment]
@mikeduminy got it. Here are my first thoughts:
- It is very unlikely that HERMES_LIMIT_STACK_DEPTH is set - it is only enabled in sanitizer builds of Hermes, which you are almost certainly not running.
- This actually looks like a JS stack overflow, not a native stack overflow.

Are you getting a stack trace when it happens and what does it look like?

**Comment by mikeduminy** [comment]
Yeah, I think you're right about the being a JS stack overflow. I can move the discussion to a different thread if needed?

This is the error I get:
```
Maximum call stack size exceeded

no stack
```

To be clear, this doesn't happen when using JSC as a runtime. Is there a way to get a stack trace? Perhaps by building hermes locally or enabling some debug flags?

**Comment by maximilize** [comment]
This error happens for example with the `ajv` module to verify JSON schemas. I checked their code and see that it would be very difficult to change the code logic to prevent many stacked function calls.

To see the maximum stack trace, this code might help to find it:

```ts
const fn = (i) => {
  console.log("i =", i);
  fn(i + 1);
};

fn(0);
```

**Comment by tmikov** [comment]
@mikeduminy something strange is going on here, this is not a "normal" stack overflow. There are a few of things that stick out to me.

**First**, Hermes has a fairly large register stack, it usually allows deeper recursion than JSC or v8. You can run this example locally and see what it prints:
```
let depth = 0;
let printStack = false;

function foo() {
    ++depth;
    foo();
}
try {
    foo();
} catch (e) {
    if (printStack)
        print(e.stack);
    else
        print("depth =", depth, e);
}
```
When I run it on my Mac, I get 95319 recursive invocations, which is pretty deep (compared to 13997 for v8, 45615 for JSC).

The **second** thing to notice is that JS exceptions in Hermes always carry a stack trace, regardless of how it is built, even in production mode. If you change `printStack` to `true` in the above example, the trace will be printed. Even if you strip the debug information from the bytecode, etc, something will always be printed. 

My **third** observation is that, AFAICT (I just performed a search), Hermes itself does not contain the string "no stack" from your error. 

The last point is particularly interesting. Something in the integration must be intercepting the stack trace and erasing it. 

As I understand it, you are running this in "developer mode", with Metro, fast refresh, etc? So, there are quite a few pieces of the puzzle there that are working together at runtime, which I actually don't understand well (I understand Hermes quite well, but not necessarily how it is integrated in React Native and Metro). We should do a search in React Native, Metro, etc, for something generating the string "no stack".

This is what I got so far, off the top of my head. After the holidays (on Tuesday), I will discuss it with others from the Hermes team to see whether they have ideas.

**Comment by mikeduminy** [comment]
Thanks for being so helpful @tmikov, I appreciate it very much.

I'll spend some time looking into what you suggested, but wanted to share that I found a potential source of the "no stack" message in [react-native JSI code](https://github.com/facebook/react-native/blob/main/ReactCommon/jsi/jsi/jsi.cpp#L513) 👀

**Comment by neildhar** [comment]
While it's likely the "no stack" message is coming from there, it's not clear why the stack is being discarded. Something else may be catching the exception and stripping some of the information. It would be interesting to poke at the exception and look at its type and prototype (if applicable), that may give some hint as to where it is coming from. (since it is missing the stack property)

You can get that value just by catching the exception in JS, or by using `JSError::value()`.

Another thing to note is that there was actually a bug in our stack configuration that caused the stack limit to be artificially low, even when a larger stack limit is specified. This was fixed in https://github.com/facebook/hermes/commit/03f2dffc1d0ef8b2360a6790ad425ce4013e4de3. It is not yet in RN, but when it is, it will allow you to raise the stack limit using `RuntimeConfig::Builder::withMaxNumRegisters`. This bug is specific to using Hermes through JSI, which is why it does not show up when you use the `hermes` CLI tool directly.

**Comment by tmikov** [comment]
@mikeduminy so, it seems like the RN+Hermes version you are using may be incorrectly configured with 16x smaller stack size. Are you able to build Hermes locally and patch one number? Change https://github.com/facebook/hermes/blob/2656b239f5d0e25ce0bdf9327718355f8f0b5e87/API/hermes/hermes.cpp#L118 to just be `= 1024 * 1024`.

**Comment by mikeduminy** [comment]
@tmikov That fixed it 😮 👍 🙇 

@neildhar Thanks for the detail. Would you like us to dig deeper into the value of the error given that the change above fixed it?

Do you have any idea what version of react-native the fix (https://github.com/facebook/hermes/commit/03f2dffc1d0ef8b2360a6790ad425ce4013e4de3) would be in?

**Comment by mikeduminy** [comment]
> @mikeduminy so, it seems like the RN+Hermes version you are using may be incorrectly configured with 16x smaller stack size. Are you able to build Hermes locally and patch one number? Change
> 
> https://github.com/facebook/hermes/blob/2656b239f5d0e25ce0bdf9327718355f8f0b5e87/API/hermes/hermes.cpp#L118
> 
> to just be `= 1024 * 1024`.

Are there any risks to shipping with this (16x) change? Is it possible that we could break things on lower-end devices?

**Comment by tmikov** [comment]
@mikeduminy before discussing shipping this to production, does it fix your problem? Also, I thought that you only saw the problem during local development (in debug mode)?

**Comment by tmikov** [comment]
@mikeduminy This value consumes 8 MB virtual address space, which is only "converted" to RAM when actually used. So, in theory it shouldn't break things on any device. However it is probably more than you need, since I believe you were seeing the problem only in dev mode locally?

Have you tried finding the minimum value that works locally (using binary search)?

If I had to ship something in a hurry, I would use 512 * 1024.

**Comment by mikeduminy** [comment]
@tmikov Gotcha 👍 

Yes, we're only seeing it locally in dev mode, but we'd like to use the same hermes build in both dev and prod to catch any problems in development. Also, on iOS swapping the code out between environments would result in changes to the pod lockfile which is annoying to have to ignore locally (because sometimes you might actually have cause to change this).

**Comment by neildhar** [comment]
@mikeduminy 

> Would you like us to dig deeper into the value of the error 

It's certainly still odd that the stack trace was deleted. If you do investigate further, I'd be curious to hear what you find. Especially since this may be a sign of a bug somewhere in Hermes/RN.

> Do you have any idea what version of react-native the fix (https://github.com/facebook/hermes/commit/03f2dffc1d0ef8b2360a6790ad425ce4013e4de3) would be in?

I expect it to be in 0.72. But I'll share this discussion with the RN team in case it makes sense to include it in a point release for 0.71 given that it is impacting users.

**Comment by tmikov** [comment]
@neildhar @mikeduminy correct me if I am wrong, but https://github.com/facebook/hermes/commit/03f2dffc1d0ef8b2360a6790ad425ce4013e4de3 doesn't actually fix this problem because it limits the stack size to 64 * 1024?

**Comment by neildhar** [comment]
@tmikov It allows you to configure the stack size via the `RuntimeConfig`, so you would be able to adjust the heap size without maintaining a custom build of Hermes.

I'm not sure how conveniently RN surfaces the `RuntimeConfig` options to developers though (and therefore whether this is necessarily an improvement). Perhaps @mikeduminy knows how easy it would be to make this change.

**Comment by tmikov** [comment]
@neildhar I believe we should change the default config to at least 256 * 1024 (pending confirmation from @mikeduminy that such a size works). I believe 512 * 1024 might be needed for parity with JSC, which, on my system, is able to make about 45,000 recursive calls (though it is possible that it is different on mobile).

**Comment by mikeduminy** [comment]
> @tmikov It allows you to configure the stack size via the `RuntimeConfig`, so you would be able to adjust the heap size without maintaining a custom build of Hermes.
> 
> I'm not sure how conveniently RN surfaces the `RuntimeConfig` options to developers though (and therefore whether this is necessarily an improvement). Perhaps @mikeduminy knows how easy it would be to make this change.

I don't know of any RuntimeConfig options being surfaced through RN but we could always apply local patches to RN to make this work. The first step is to make sure this ability makes its way to RN. 

Ideally we'd have this change available in 0.70 (the version we are on), or 0.71 (the latest version), instead of 0.72 (the next version). I took a look at how RN builds and bundles Hermes and I'm not sure if it would be possible to easily do since hermes doesn't follow a tagged/branched version strategy, but I'm curious if you think it would be possible to backport this change to either of those versions?

**Comment by mikeduminy** [comment]
Is a custom patch for previous RN versions something that the Hermes team is willing to do? @tmikov @neildhar

**Comment by tmikov** [comment]
@mikeduminy we can probably release patches setting the correct default stack size: 128 or 256. I am saying "probably", since I am not familiar with the RN release process. cc: @neildhar who understand it much better.

**Comment by mikeduminy** [comment]
@tmikov thanks, I want to first confirm something about the code before the change.
https://github.com/facebook/hermes/blob/2656b239f5d0e25ce0bdf9327718355f8f0b5e87/API/hermes/hermes.cpp#L118-L120

Can you help me understand what the values of `sizeof(::hermes::vm::Runtime)` and `sizeof(::hermes::vm::PinnedHermesValue)` would be?

Earlier you mentioned that 1024*1024 was 16x more than we had before, which implies this value was 64*1024 before. However, in my testing, 64*1024 seems to work (32*1024 does not). I'd like to make sure that my results make sense before going further, as I could have made a mistake during testing.

**Comment by tmikov** [comment]
@mikeduminy this is just a hack to calculate the size of the Hermes register stack that can fit with the runtime on a 512KB native thread stack. So we have the size of the native thread stack (512KB) minus the size of the runtime, minus 8 4KB pages of extra space, and we divide that by the size of a Hermes register, to get the available register stack size.

All of that is outdated and should be removed.

I don't actually know what it calculates to, I just assumed that it was 64K, but I may be way off. 

128 * 1024 registers (which is 128 * 1024 * 8 bytes) seems reasonable to me.

**Comment by neildhar** [comment]
> I'd like to make sure that my results make sense before going further, as I could have made a mistake during testing

Technically it was slightly lower than `64 * 1024` because we subtracted some extra space as tmikov mentioned above. That extra space is likely just enough for your application to run.

> custom patch for previous RN versions

Those tags are maintained and used by the RN team. If the RN team is planning further point releases for 0.70 and 0.71, the fix should port cleanly to them. cc @cortinico regarding backporting

**Comment by cortinico** [comment]
Thanks for the heads up. We can totally backport Hermes changes inside RN (the infras was set up to allow for this).

If you're interested in having anything backported to a release series of RN, please comment here: https://github.com/reactwg/react-native-releases/discussions

(Provided https://github.com/facebook/hermes/commit/03f2dffc1d0ef8b2360a6790ad425ce4013e4de3 is safe to cherry-pick on top of previous version of Hermes).

Anyway, as of today, this change is scheduled to land in the Bundled Hermes with React Native 0.72

**Comment by tmikov** [comment]
I just want to remind everybody, again, that https://github.com/facebook/hermes/commit/03f2dffc1d0ef8b2360a6790ad425ce4013e4de3 does not fix this problem. We need another change on top of it, for the next RN release, and separate patches for previous releases.

**Comment by mikeduminy** [comment]
> I just want to remind everybody, again, that [03f2dff](https://github.com/facebook/hermes/commit/03f2dffc1d0ef8b2360a6790ad425ce4013e4de3) does not really fix this problem. We need another change on top of it, for the next RN release, and separate patches for previous releases.

So you are suggesting a bump to 128? I can make that change if everyone agrees to it. 

Beyond that the integration into RN should probably expose a way in userland to adjust this limit. I see several points in RN where the RuntimeConfig is created and used - places where `withMaxNumRegisters` could be called to change the limit. That change is a bigger one and needs more specific knowledge about where to make the changes. So for now I vote for the proposed change to `128*1024` and backporting at least to 0.71, maybe to 0.70.

**Comment by tmikov** [comment]
@mikeduminy I agree.

128*1024 should work nicely. It is only 1 MB of address space, and the actual RAM is consumed only if the stack gets that high, so we are not wasting memory.

In older versions we need to update `kMaxNumRegisters`, but in the current version (after https://github.com/facebook/hermes/commit/03f2dffc1d0ef8b2360a6790ad425ce4013e4de3), we need to update the default configuration size here: https://github.com/facebook/hermes/commit/03f2dffc1d0ef8b2360a6790ad425ce4013e4de3#diff-fe65dc57e9fc0262272f68b4e66424c75dfe8a4a2cf5d02bffdfc4aa8d9aefecR38 . 

Then RN won't ideally need to call withMaxNumRegisters (it is probably a good idea to allow customization in some way, but that is an orthogonal problem to the default stack size being too low in Hermes).

**Comment by mikeduminy** [comment]
PRs have been created to double the default in Hermes and apply the same default to the 0.71 and 0.70 tracking branches. I'm coordinating with the RN team via the [0.71.4 discussion](https://github.com/reactwg/react-native-releases/discussions/58#discussioncomment-5122408) and the [0.70.8 discussion](https://github.com/reactwg/react-native-releases/discussions/56#discussioncomment-5122392). What would really help move these over the line is for the hermes team to review (and hopefully approve) the linked PRs.

If I've missed anything please let me know 🙏

**Comment by neildhar** [comment]
@mikeduminy thanks for making those. I've imported your PR to the main branch. Once that lands, I'll defer to the RN team on how to transplant it onto the release branches, the other PRs are probably not needed.

**Comment by mikeduminy** [comment]
Happy to help, and thank you to all involved (especially @tmikov) for helping us diagnose and fix this edge case 🙏 

For the RN team, the PRs for [0.70](https://github.com/facebook/hermes/pull/924) and [0.71](https://github.com/facebook/hermes/pull/925) show the compatible change for those respective versions and are really what will make this change useful for us today.

**Comment by kelset** [comment]
> thanks for making those. I've imported your PR to the main branch. Once that lands, I'll defer to the RN team on how to transplant it onto the release branches, the other PRs are probably not needed.

@neildhar I think we'll actually need to merge the PRs from @mikeduminy since the change for main is different from the ones for 0.71/0.70.

**Comment by neildhar** [comment]
@kelset I think we could just pick both 03f2dffc1d0ef8b2360a6790ad425ce4013e4de3 and https://github.com/facebook/hermes/pull/923 (once it is committed). That minimises divergence from trunk, and fixes the underlying bug in setting the register stack size via the `RuntimeConfig`, in the event that someone does want to build from source and specify `withMaxNumRegisters`.

**Comment by kelset** [comment]
> @kelset Lorenzo Sciandra FTE I think we could just pick both [03f2dff](https://github.com/facebook/hermes/commit/03f2dffc1d0ef8b2360a6790ad425ce4013e4de3) and #923 (once it is committed). That minimises divergence from trunk, and fixes the underlying bug in setting the register stack size via the `RuntimeConfig`, in the event that someone does want to build from source and specify `withMaxNumRegisters`.

sounds good to me - in that case then the other two PRs from @mikeduminy should be closed.

**Comment by Waltari10** [comment]
Hi, I'm trying to upgrade to Hermes in our RN app and I'm getting the `ERROR  RangeError: Maximum call stack size exceeded (native stack depth), js engine: hermes`. No big JSON files.

Is the default max stack size going to be increased in RN 0.72?

**Comment by mikeduminy** [comment]
@Waltari10 the fix has landed in [RN 0.71.4](https://github.com/facebook/react-native/releases/tag/v0.71.4) and will likely be cherry picked into [RN 0.70.9](https://github.com/reactwg/react-native-releases/discussions/62). It will also be in RN 0.72.

**Comment by Waltari10** [comment]
> @Waltari10 the fix has landed in [RN 0.71.4](https://github.com/facebook/react-native/releases/tag/v0.71.4) and will likely be cherry picked into [RN 0.70.9](https://github.com/reactwg/react-native-releases/discussions/62). It will also be in RN 0.72.

Thanks for the info! Curiously we are already using 0.71.4 in our app though and still getting the error 🤔

**Comment by mikeduminy** [comment]
Disclaimer: I'm not an expert.

Oh I just noticed that your error message includes `(native stack depth)` which is a different error. The one that was fixed in RN 0.71.4 was a JS Register max stack size error, but yours is about the native stack size (which is much lower). You can find more information in this [comment](https://github.com/facebook/hermes/issues/135#issuecomment-1026201926).

> No big JSON files. I guess this is just Javascript code accessing the bridge too often?
Since this is a native stack issue I guess that neither JSON nor frequent bridge access is the problem. What does your stack trace look like?

Have a look at the stack trace in [this comment](https://github.com/facebook/hermes/issues/956#issuecomment-1494774575). Notice the multiple lines that say: `at apply (native)`. This indicates that a native call is being made. Note: this is not a react native call, but rather what the runtime considers a native call, which would be things like [Function.prototype.call](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/call), [Function.prototype.apply](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply), or (maybe) [Array.prototype.map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map).

Maybe the transpilation step of your builds is adding too many of these native calls to functions that are called in a deeply-nested way. You could see if any changes to your babel settings could fix it. Or perhaps you need to hunt down what changed in your codebase from when it was working.

**Comment by aurangs7** [comment]
> > This happened to me when I upgraded my react-native https://react-native-community.github.io/upgrade-helper/?from=0.63.4&to=0.64.0
> > ```
> >  WARN  Require cycle: node_modules/core-js/internals/microtask.js -> node_modules/core-js/internals/microtask.js
> > 
> > Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
> >  WARN  Require cycle: node_modules/rn-fetch-blob/index.js -> node_modules/rn-fetch-blob/polyfill/index.js -> node_modules/rn-fetch-blob/polyfill/Blob.js -> node_modules/rn-fetch-blob/index.js
> > 
> > Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
> >  WARN  Require cycle: node_modules/rn-fetch-blob/index.js -> node_modules/rn-fetch-blob/polyfill/index.js -> node_modules/rn-fetch-blob/polyfill/XMLHttpRequest.js -> node_modules/rn-fetch-blob/index.js
> > 
> > Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
> >  WARN  Require cycle: node_modules/rn-fetch-blob/index.js -> node_modules/rn-fetch-blob/polyfill/index.js -> node_modules/rn-fetch-blob/polyfill/Fetch.js -> node_modules/rn-fetch-blob/index.js
> > 
> > Require cycles are allowed, but can result in uninitialized values. Consider refactoring to remove the need for a cycle.
> >  ERROR  RangeError: Maximum call stack size exceeded, js engine: hermes
> >  LOG  Running "ny_app" with {"rootTag":1}
> >  ERROR  Invariant Violation: "ny_app" has not been registered. This can happen if:
> > * Metro (the local dev server) is run from the wrong folder. Check if Metro is running, stop it and restart it in the current project.
> > * A module failed to load due to an error and `AppRegistry.registerComponent` wasn't called., js engine: hermes
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Happens both on android and ios.
> 
> Setting inlineRequires: false in metro.config.js fixed it for me

I am also getting this error in Android (iOS working fine), this happened suddenly. 
It was working fine till yesterday. 

  inlineRequires: false => is already added in file

Any solution? 

![Screenshot_1683805028](https://github.com/facebook/hermes/assets/110371514/226f4ab5-3370-43ad-bf88-fce4458ff62b)

**Comment by antondalgren** [comment]
Hi,

Been struck by this issue too when migrating from RN 0.69 to 0.70 and enabling hermes in the process.

Would this limitation of the call stack on `Function.call()` also be applicable to the usage of `Array.prototype.reduce`? As the implementation of the reduce function is calling the `Callable::executeCall4` which in turn uses `ScopedNativeCallFrame`. 

https://github.com/facebook/hermes/blob/d1be3ff5cabf16e96615b97e057d641143a11b51/lib/VM/JSLib/Array.cpp#L3276
https://github.com/facebook/hermes/blob/081b8ac39eccc09ed67a6363649d5ed21ceda8ef/lib/VM/Callable.cpp#L296

We are essentially using AJV for JSON validation, and that library is using recursion within the `call` and `reduce` method.

Also noted that `Function.apply()` is also limited by this call stack.
https://github.com/facebook/hermes/blob/d1be3ff5cabf16e96615b97e057d641143a11b51/lib/VM/JSLib/Function.cpp#L211

**Comment by tmikov** [comment]
FWIW, this has been addressed in the next version of Hermes, but unfortunately we do not yet have a timeline for release or for backporting.

**Comment by vmalvaro** [comment]
Hi @tmikov,

Would you mind helping me to clarify some things I'm a bit confused about?

I just got hit by the `RangeError: Maximum call stack size exceeded (native stack depth)` issue while working on my app.

This seems to be different from the original issue of this thread that was fixed in https://github.com/facebook/hermes/issues/135#issuecomment-1506616481
I'm correct? fixing this issue will cover both cases or we should have another issue to track the "native stack depth" one?

In any case, my issue is that I'm using a library that needs to do recursive work where it uses apply/call functions multiple times.
I understand that the functions are converted to native calls and there is stack limit of 128. On certain situations I hit the limit and got the error.

It works fine if I use JavaScriptCore as the engine, I did a rough test in the stack limit seems to be around ~2738 when using JSC.

Is there a way to increase the limit when using Hermes or any other workaround?
Unfortunately, I can't change a 3rd party library that my app depends on.

Here is a repo with issue: https://github.com/vmalvaro/hermes-range-error

Thanks!

**Comment by tmikov** [comment]
@vmalvaro the limit can be increased by changing Hermes as described in this [comment](https://github.com/facebook/hermes/issues/135?fbclid=IwAR0mFIn7sED9QUhwY9FSqpS65ViDmKXU74Z42m1_flirtqrD2Og8TybxTKM#issuecomment-1026220296).

**Comment by vmalvaro** [comment]
@tmikov thanks for the quick response, I'm a bit of a newbie with Hermes so I'd appreciate your guidance.

This https://github.com/facebook/hermes/issues/135?fbclid=IwAR0mFIn7sED9QUhwY9FSqpS65ViDmKXU74Z42m1_flirtqrD2Og8TybxTKM#issuecomment-1026220296 means that I have to do that change in [hermes/include/hermes/VM/Runtime.h](https://github.com/facebook/hermes/blob/2c66a23882bb7b6c9f80fadd84df643e73bbd5ee/include/hermes/VM/Runtime.h#L905), then a custom build of Hermes + react native from source as described here https://hermesengine.dev/docs/react-native-integration?

**Comment by neildhar** [comment]
@vmalvaro 

> custom build of Hermes + react native from source as described here

Those instructions are unfortunately no longer accurate for newer versions of RN, which have changed how Hermes is consumed. If you're on an RN version that is 0.69+, you can set an environment variable to point RN to the source directory where you want it to get Hermes from. See https://github.com/facebook/react-native/blob/fd9e295befcd8781190ec26a6a2fc4ef39fb1c15/packages/react-native/ReactAndroid/hermes-engine/build.gradle#L42.

**Comment by Villar74** [comment]
![screen](https://github.com/facebook/hermes/assets/27006310/7bfc5c46-6b7f-411a-b9a4-79118695016c)

Get this issue when disabling js DEV mode 

```bash
 System:
    OS: macOS 13.4
    CPU: (12) arm64 Apple M2 Pro
  Binaries:
    Node: 20.0.0 - /opt/homebrew/bin/node
    Yarn: 1.22.19 - /opt/homebrew/bin/yarn
    npm: 9.6.4 - /opt/homebrew/bin/npm
    Watchman: 2023.05.01.00 - /opt/homebrew/bin/watchman
  SDKs:
    iOS SDK:
      Platforms: DriverKit 22.4, iOS 16.4, macOS 13.3, tvOS 16.4, watchOS 9.4
    Android SDK:
      API Levels: 31, 33
      Build Tools: 30.0.2, 30.0.3, 31.0.0, 33.0.0, 33.0.2
      System Images: android-33 | Google APIs ARM 64 v8a
  IDEs:
    Android Studio: Flamingo 2022.2.1 Patch 1 Flamingo 2022.2.1 Patch 1
    Xcode: 14.3/14E222b - /usr/bin/xcodebuild
  npmPackages:
    react: 18.2.0 => 18.2.0 
    react-native: 0.71.8 => 0.71.8 
```

**Comment by tmikov** [comment]
@Villar74 is your issue related to loading JSON, which is the topic of this specific issue? 

If not, and you believe there is a bug in Hermes, please create a new issue, preferably with with more details, including source reproducing the problem. It is almost impossible to diagnose a problem from a screenshot of a stack trace of a compiled bundle.

**Comment by tmikov** [comment]
Finally, this long standing problem has been fixed in https://github.com/facebook/hermes/commit/3213794cba3cb21a901da4a36a7396a5889b5481 and should be available in the next release of RN.

**Comment by Yashtoddle** [comment]
Hey tmikov could you please  send this update in the coming patch as I needed this issue to be resolved in our codebase

**Comment by tmikov** [comment]
@Yashtoddle I don't understand what you mean. Send what where?

**Comment by tmikov** [comment]
@Yashtoddle we don't really control which changes are picked into each patch releases. You also need to be more specific about the version where you want this. 

In any case I think you can request a pick here: https://github.com/reactwg/react-native-releases/discussions .

This is not a trivial bugfix, it is several commits including 18aa617c28709281be62c3b84535afa982d1feef .

@neildhar @cortinico what is the feasibility of picking this?

**Comment by neildhar** [comment]
@Yashtoddle this will be included in the upcoming RN 0.73. Which RN release are you currently on? The feasibility of picking this fix will likely depend heavily on that.
